File: android/app/src/main/java/com/nextpage/data/remote/catalog/CompositeCatalogProvider.kt
package com.nextpage.data.remote.catalog

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import com.nextpage.debug.SentryMetrics
import kotlinx.serialization.json.Json

/**
 * Ordered dynamic composite behind the port.
 * Providers are searched in construction order (built-ins first, curated,
 * then enabled addons in install order); results concat-merge in that order.
 * Page/details caching is per-source with v2 keys carrying the full source
 * string, and reads enforce the existence check (design A1): entries whose
 * source id is not in the active source set at read time are never served.
 * Search entry is direct (cache pre-read then executeSearch); the single
 * 250ms trailing debounce is owned by DiscoverViewModel (U2).
 * Mirrors desktop `CompositeCatalogProvider.ts`.
 */
class CompositeCatalogProvider(
    private val providers: List<CatalogProvider>,
    private val cache: DiscoverCacheStore? = null,
    private val nowEpochSecs: () -> Long = { System.currentTimeMillis() / 1000 }
) : CatalogProvider {

    private val cacheJson = Json { ignoreUnknownKeys = true }

    /** Sources in provider order, deduped by sourceId (first occurrence wins). */
    override fun listSources(): List<CatalogSourceInfo> {
        val seen = mutableSetOf<String>()
        val out = mutableListOf<CatalogSourceInfo>()
        for (provider in searchableProviders()) {
            for (source in provider.listSources()) {
                if (seen.add(source.sourceId)) out.add(source)
            }
        }
        return out
    }

    /**
     * Direct entry point (U2): cache pre-read, then straight to executeSearch.
     * Fresh cache entries return immediately; page < 1 rejects before I/O.
     */
    override suspend fun search(query: String, page: Int): PagedResult {
        if (page < 1) throw catalogError(CatalogErrorCode.INVALID_PAGE, "page must be >= 1, got $page")
        readAllProviderPages(query, page)?.let { return it }
        return executeSearch(query, page)
    }

    private suspend fun executeSearch(query: String, page: Int): PagedResult = coroutineScope {
        // Re-check before fan-out: a concurrent caller may have filled the cache.
        readAllProviderPages(query, page)?.let { return@coroutineScope it }
        val active = activeSourceIds()
        val searches = searchableProviders().map { provider ->
            async {
                val startedMs = System.currentTimeMillis()
                try {
                    val cached = readProviderPage(provider, query, page, active)
                    val sourceId = singleSource(provider)?.sourceId ?: "unknown"
                    if (cached != null) {
                        emitDiscoverSearch(cached, sourceId, startedMs, true)
                        cached
                    } else {
                        val result = provider.search(query, page)
                        cacheProviderPage(provider, query, page, result, active)
                        emitDiscoverSearch(result, sourceId, startedMs, false)
                        result
                    }
                } catch (err: Throwable) {
                    // Failure isolation (U1): one failing source fails closed to
                    // an empty page so it can never fail the whole fan-out.
                    // Cancellation still propagates to respect coroutine scope.
                    if (err is CancellationException) throw err
                    val code = (err as? CatalogException)?.code?.name
                        ?: CatalogErrorCode.UPSTREAM_ERROR.name
                    val sourceId = singleSource(provider)?.sourceId ?: "unknown"
                    SentryMetrics.count(
                        "discover_search_error",
                        mapOf("provider" to sourceId, "code" to code)
                    )
                    PagedResult(emptyList(), null, 0)
                }
            }
            }
            mergePaged(searches.map { it.await() }, page)
    }

    /**
     * Whole-composite cache pre-read: when every searchable provider has a
     * fresh page entry, the merged page returns without debounce or I/O.
     */
    private suspend fun readAllProviderPages(query: String, page: Int): PagedResult? {
        val cache = cache ?: return null
        val active = activeSourceIds()
        val providers = searchableProviders().filter { it.listSources().size == 1 }
        if (providers.isEmpty()) return null
        val pages = mutableListOf<PagedResult>()
        for (provider in providers) {
            val cached = readProviderPage(provider, query, page, active) ?: return null
            pages.add(cached)
        }
        return mergePaged(pages, page)
    }

    /** Providers without sources (disabled) contribute nothing and are never called. */
    private fun searchableProviders(): List<CatalogProvider> =
        providers.filter { it.listSources().isNotEmpty() }

    private fun activeSourceIds(): Set<String> =
        searchableProviders().flatMap { it.listSources() }.map { it.sourceId }.toSet()

    /** Existence check (design A1): only sources active at read time may hit. */
    private suspend fun readProviderPage(
        provider: CatalogProvider,
        query: String,
        page: Int,
        active: Set<String>
    ): PagedResult? {
        val store = cache ?: return null
        val source = singleSource(provider) ?: return null
        if (source.sourceId !in active) return null
        val hit = store.get(pageCacheKey(source.sourceId, query, page), nowEpochSecs()) ?: return null
        return cacheJson.decodeFromString(PagedResult.serializer(), hit)
    }

    private suspend fun cacheProviderPage(
        provider: CatalogProvider,
        query: String,
        page: Int,
        result: PagedResult,
        active: Set<String>
    ) {
        // Never-cache-empty: empty search results persist nothing and extend
        // no TTL, so a prior valid cached page keeps its existing TTL.
        if (result.results.isEmpty()) return
        val store = cache ?: return
        val source = singleSource(provider) ?: return
        if (source.sourceId !in active) return
        store.put(
            pageCacheKey(source.sourceId, query, page),
            cacheJson.encodeToString(PagedResult.serializer(), result),
            nowEpochSecs(),
            PAGE_TTL_S
        )
    }

    /**
     * Featured page cache read for a single-source provider under the `f:v3:`
     * namespace. Same existence check as [readProviderPage]: sources that are not
     * active at read time are never served.
     */
    private suspend fun readProviderFeaturedPage(
        provider: CatalogProvider,
        sort: CatalogFeaturedSort,
        page: Int,
        active: Set<String>
    ): PagedResult? {
        val store = cache ?: return null
        val source = singleSource(provider) ?: return null
        if (source.sourceId !in active) return null
        val hit = store.get(featuredCacheKey(source.sourceId, sort, page), nowEpochSecs()) ?: return null
        return cacheJson.decodeFromString(PagedResult.serializer(), hit)
    }

    private suspend fun cacheProviderFeaturedPage(
        provider: CatalogProvider,
        sort: CatalogFeaturedSort,
        page: Int,
        result: PagedResult,
        active: Set<String>
    ) {
        // Never-cache-empty: empty rails persist nothing and extend no TTL,
        // so a prior valid cached page keeps its existing TTL.
        if (result.results.isEmpty()) return
        val store = cache ?: return
        val source = singleSource(provider) ?: return
        if (source.sourceId !in active) return
        store.put(
            featuredCacheKey(source.sourceId, sort, page),
            cacheJson.encodeToString(PagedResult.serializer(), result),
            nowEpochSecs(),
            FEATURED_TTL_S
        )
    }

    /**
     * Ordered merge: left-fold the provider pages — earlier providers win fields,
     * later ones fill cover gaps and append unmatched books (the [Gutendex,
     * OpenLibrary] fold reproduces the legacy hardcoded-pair merge exactly).
     */
    /**
     * U3-2 search emission: latency distributionRaw + request/empty
     * counters per provider completion. NEVER emits user IDs, query
     * text, or book IDs; attribute set stays {provider, cached} /
     * {surface, provider}.
     */
    private fun emitDiscoverSearch(
        page: PagedResult,
        sourceId: String,
        startedMs: Long,
        cached: Boolean
    ) {
        val elapsedMs = System.currentTimeMillis() - startedMs
        SentryMetrics.distributionRaw(
            "discover_search_latency",
            elapsedMs,
            mapOf("provider" to sourceId, "cached" to cached.toString())
        )
        SentryMetrics.count(
            "discover_request_total",
            mapOf("surface" to "search", "provider" to sourceId)
        )
        if (page.results.isEmpty()) {
            SentryMetrics.count(
                "discover_empty_total",
                mapOf("surface" to "search", "provider" to sourceId)
            )
        }
    }

    /**
     * U3-2 rail emission: latency distributionRaw + request/empty
     * counters per provider completion. NEVER emits user IDs, query
     * text, or book IDs; attribute set stays {provider, cached} /
     * {surface, provider}.
     */
    private fun emitDiscoverRail(
        page: PagedResult,
        sourceId: String,
        startedMs: Long,
        cached: Boolean
    ) {
        val elapsedMs = System.currentTimeMillis() - startedMs
        SentryMetrics.distributionRaw(
            "discover_search_latency",
            elapsedMs,
            mapOf("provider" to sourceId, "cached" to cached.toString())
        )
        SentryMetrics.count(
            "discover_request_total",
            mapOf("surface" to "rail", "provider" to sourceId)
        )
        if (page.results.isEmpty()) {
            SentryMetrics.count(
                "discover_empty_total",
                mapOf("surface" to "rail", "provider" to sourceId)
            )
        }
    }

    private fun mergePaged(pages: List<PagedResult>, page: Int): PagedResult {
        if (pages.isEmpty()) return toPagedResult(emptyList(), page, 0)
        var results = pages.first().results
        var totalCount = pages.first().totalCount
        for (i in 1 until pages.size) {
            results = mergeResults(results, pages[i].results)
            totalCount = resolveTotalCount(totalCount, pages[i].totalCount)
        }
        return toPagedResult(results, page, totalCount)
    }

    /**
     * True when at least one active provider opts in, so the shell can decide
     * whether any featured work is possible at all.
     */
    override fun supportsFeatured(): Boolean =
        searchableProviders().any { it.supportsFeatured() }

    /**
     * Featured rails fan out with `async` over the providers that opt in via
     * [CatalogProvider.supportsFeatured], then merge with the same [mergePaged]
     * left-fold used by search. A provider that does not opt in is never called,
     * so its rail can only ever come back empty (fail-closed) and be hidden.
     */
    override suspend fun featured(sort: CatalogFeaturedSort, page: Int): PagedResult {
        if (page < 1) throw catalogError(CatalogErrorCode.INVALID_PAGE, "page must be >= 1, got $page")
        return coroutineScope {
            val active = activeSourceIds()
            val pages = searchableProviders()
                    .filter { it.supportsFeatured() }
                    .map { provider ->
                        async {
                            val startedMs = System.currentTimeMillis()
                            try {
                                val cached = readProviderFeaturedPage(provider, sort, page, active)
                                val sourceId = singleSource(provider)?.sourceId ?: "unknown"
                                if (cached != null) {
                                    emitDiscoverRail(cached, sourceId, startedMs, true)
                                    cached
                                } else {
                                    val result = provider.featured(sort, page)
                                    cacheProviderFeaturedPage(provider, sort, page, result, active)
                                    emitDiscoverRail(result, sourceId, startedMs, false)
                                    result
                                }
                            } catch (err: Throwable) {
                                if (err is CancellationException) throw err
                                currentCoroutineContext().ensureActive()
                                // U3-2 (U1-3 catch path): error counter is the ONLY
                                // addition authorized here; isolation stays untouched.
                                val code = (err as? CatalogException)?.code?.name
                                    ?: CatalogErrorCode.UPSTREAM_ERROR.name
                                val sourceId = singleSource(provider)?.sourceId ?: "unknown"
                                SentryMetrics.count(
                                    "discover_search_error",
                                    mapOf("provider" to sourceId, "code" to code)
                                )
                                PagedResult(emptyList(), null, 0)
                            }
                        }
                    }
                mergePaged(pages.map { it.await() }, page)
        }
    }

    /**
     * Per-source search: exact match over the active source set, routed to the
     * single provider that owns [sourceId]. An unknown or inactive source fails
     * closed with an empty page — never a crash, never a silent composite search.
     */
    override suspend fun searchSource(sourceId: String, query: String, page: Int): PagedResult {
        if (page < 1) throw catalogError(CatalogErrorCode.INVALID_PAGE, "page must be >= 1, got $page")
        val owner = searchableProviders().firstOrNull { provider ->
            provider.listSources().any { it.sourceId == sourceId }
        } ?: return PagedResult(emptyList(), null, 0)
        return owner.search(query, page)
    }

    /**
     * Exact-prefix routing over the active source set: `gutendex:`/`openlibrary:`
     * book ids hit the built-ins, `addon:<addonId>:<bookId>` the owning addon.
     * Unroutable ids reject NOT_FOUND without any I/O.
     */
    override suspend fun getDetails(id: String): CatalogBook {
        val route = routeDetails(id)
            ?: throw catalogError(CatalogErrorCode.NOT_FOUND, "unknown catalog id $id")
        val active = activeSourceIds()
        if (cache != null && route.source.sourceId in active) {
            val hit = cache.get(detailCacheKey(route.source.sourceId, id), nowEpochSecs())
            if (hit != null) return cacheJson.decodeFromString(CatalogBook.serializer(), hit)
        }
        val book = route.provider.getDetails(id)
        if (cache != null && route.source.sourceId in active) {
            cache.put(
                detailCacheKey(route.source.sourceId, id),
                cacheJson.encodeToString(CatalogBook.serializer(), book),
                nowEpochSecs(),
                DETAIL_TTL_S
            )
        }
        return book
    }

    private fun routeDetails(id: String): RoutedDetails? {
        var best: RoutedDetails? = null
        var bestPrefix = ""
        for (provider in searchableProviders()) {
            for (source in provider.listSources()) {
                val prefix = bookIdPrefixForSource(source.sourceId) ?: continue
                if (!id.startsWith(prefix) || id.length <= prefix.length) continue
                if (best == null || prefix.length > bestPrefix.length) {
                    best = RoutedDetails(provider, source)
                    bestPrefix = prefix
                }
            }
        }
        return best
    }

    override fun resolveDownloadUrl(formats: Map<String, String>, preferEpub: Boolean): String =
        com.nextpage.data.remote.catalog.resolveDownloadUrl(formats, preferEpub)

    data class RoutedDetails(val provider: CatalogProvider, val source: CatalogSourceInfo)

    companion object {
        /** Providers exposing exactly one source are page-cacheable under that source. */
        fun singleSource(provider: CatalogProvider): CatalogSourceInfo? =
            provider.listSources().singleOrNull()

        /**
         * Book-id prefix owned by a source: built-ins drop the `builtin:` namespace
         * (`builtin:gutendex` -> `gutendex:`); addons use the source id itself
         * (`addon:<id>` -> `addon:<id>:`).
         */
        fun bookIdPrefixForSource(sourceId: String): String? = when {
            sourceId.startsWith("builtin:") -> "${sourceId.removePrefix("builtin:")}:"
            sourceId.startsWith("addon:") -> "$sourceId:"
            else -> null
        }
    }
}

File: desktop/src/lib/shared/services/catalog/CompositeCatalogProvider.ts
/**
 * CompositeCatalogProvider — ordered dynamic composite behind the port.
 * Providers are searched in construction order (built-ins first, curated,
 * then enabled addons in install order); results concat-merge in that order.
 * Page/details caching is per-source with v2 keys carrying the full source
 * string, and reads enforce the existence check (design A1): entries whose
 * source id is not in the active source set at read time are never served.
 * Burst searches are trailing-edge debounced; page < 1 rejects before I/O.
 */
import { catalogError } from './errors';
import type {
  CatalogBook,
  CatalogFeaturedSort,
  CatalogProvider,
  CatalogSource,
  CatalogSourceInfo,
  PagedResult,
} from './CatalogProvider';
import {
  DETAIL_TTL_S,
  FEATURED_TTL_S,
  PAGE_TTL_S,
  detailCacheKey,
  featuredCacheKey,
  pageCacheKey,
  type DiscoverCacheStore,
  type PreloadableCache,
} from './DiscoverCache';
import { GutendexCatalogProvider } from './BuiltInCatalogProviders';
import { GoogleBooksCatalogProvider, googleBooksProviderOrNull } from './BuiltInCatalogProviders';
import { OpenLibraryCatalogProvider } from './BuiltInCatalogProviders';
import { mergeResults, resolveDownloadUrl, resolveTotalCount, toPagedResult } from './mappers';
import { DEBOUNCE_MS, createSearchDebouncer } from './policy';

export interface CompositeOptions {
  debounceMs?: number;
  cache?: DiscoverCacheStore | null;
  nowEpochSecs?: () => number;
}

/**
 * Book-id prefix owned by a source: built-ins drop the `builtin:` namespace
 * (`builtin:gutendex` -> `gutendex:`); addons use the source id itself
 * (`addon:<id>` -> `addon:<id>:`).
 */
export function bookIdPrefixForSource(sourceId: string): string | null {
  if (sourceId.startsWith('builtin:')) return `${sourceId.slice('builtin:'.length)}:`;
  if (sourceId.startsWith('addon:')) return `${sourceId}:`;
  return null;
}

/** Providers exposing exactly one source are page-cacheable under that source. */
function singleSource(provider: CatalogProvider): CatalogSourceInfo | null {
  const sources = provider.listSources();
  return sources.length === 1 ? sources[0] : null;
}

/**
 * Resident payload read: uses the store's additive `read()` when present
 * (fresh-or-stale, non-mutating) and falls back to the fresh-only `get()` for
 * stores that only implement the original contract. A throwing or half-broken
 * store is treated as a miss — a cache must never fail a fetch.
 */
function readCachedPayload(
  store: DiscoverCacheStore,
  key: string,
  nowEpochSecs: number,
): { payload: string; stale: boolean } | null {
  const read = store.read;
  try {
    if (read) {
      const hit = read.call(store, key, nowEpochSecs);
      return hit ? { payload: hit.payload, stale: hit.stale } : null;
    }
    const fresh = store.get(key, nowEpochSecs);
    return fresh === null ? null : { payload: fresh, stale: false };
  } catch {
    return null;
  }
}

function isPreloadable(
  cache: DiscoverCacheStore | null,
): cache is DiscoverCacheStore & PreloadableCache {
  return typeof (cache as Partial<PreloadableCache> | null)?.preload === 'function';
}

import { CuratedCatalogProvider } from '../addons/CuratedCatalogProvider';
import { AddonCatalogProvider, EMPTY_ADDON_ACCESS } from '../addons/AddonCatalogProvider';
import type { AddonAccessResolution } from '../addons/AddonCatalogProvider';
import type { AddonConsentGate } from '../addons/AddonConsent';
import type { AddonTransport, InstalledAddonRow } from '../addons/AddonRegistry';

/**
 * Default provider list: built-ins first (Gutendex, Open Library, then the
 * key-gated Google Books provider), then the curated bundle, then enabled
 * addons in install order (disabled rows are excluded).
 *
 * `googleBooksKey` defaults to blank, which omits Google Books entirely
 * (fail-closed); the app composition root passes
 * `googleBooksKeyFromEnv()` so the ambient environment never leaks into
 * provider construction made by tests.
 *
 * `consent` is the per-addon network-consent gate handed to every addon
 * provider (resolve-only gating; search/getDetails stay ungated). Omitted ⇒
 * each provider denies every resolve (fail-closed default).
 */
export function defaultCatalogProviders(
  installedAddons: InstalledAddonRow[] = [],
  addonTransport?: AddonTransport,
  googleBooksKey = '',
  consent?: AddonConsentGate,
): CatalogProvider[] {
  const addonProviders = installedAddons
    .filter((row) => row.enabled)
    .map(
      (row) => new AddonCatalogProvider(row.manifest, row.id, addonTransport, undefined, consent),
    );
  const googleBooks: GoogleBooksCatalogProvider | null = googleBooksProviderOrNull(googleBooksKey);
  return [
    new GutendexCatalogProvider(),
    new OpenLibraryCatalogProvider(),
    ...(googleBooks ? [googleBooks] : []),
    new CuratedCatalogProvider(),
    ...addonProviders,
  ];
}

/**
 * Live composite supplier: rebuilds the ordered provider list from fresh
 * registry rows whenever it is invalidated (design A1/A4 wiring — install,
 * enable/disable, and uninstall immediately change the active source set).
 */
export interface CatalogProviderSupplier {
  /** Current composite; rebuilds from registry rows after invalidate(). */
  current(): Promise<CompositeCatalogProvider>;
  /** Already-built composite, or null before the first build. */
  peek(): CompositeCatalogProvider | null;
  /** Drop the cached composite; the next current() rebuilds from fresh rows. */
  invalidate(): void;
}

/** Rebuild-time options: the Discover cache the composite reads and writes. */
export interface RebuildingCatalogProviderOptions {
  cache?: DiscoverCacheStore | null;
  /** Per-addon network-consent gate forwarded to every addon provider. */
  consent?: AddonConsentGate;
}

/**
 * Seeds the durable cache mirror once per composite build, bounded to the
 * featured keys of the active sources. Absent or non-preloadable caches are a
 * no-op, and a failed preload degrades to an unseeded mirror (the first rail
 * simply refetches) — never a build failure.
 */
async function preloadDiscoverCache(
  cache: DiscoverCacheStore | null,
  composite: CompositeCatalogProvider,
): Promise<void> {
  if (!isPreloadable(cache)) return;
  try {
    await cache.preload(composite.listSources().map((source) => source.sourceId));
  } catch {
    // Best-effort: an unseeded mirror is an empty cache, not an error.
  }
}

export function createRebuildingCatalogProvider(
  loadRows: () => Promise<InstalledAddonRow[]>,
  addonTransport?: AddonTransport,
  googleBooksKey = '',
  options: RebuildingCatalogProviderOptions = {},
): CatalogProviderSupplier {
  const cache = options.cache ?? null;
  const consent = options.consent;
  let current: Promise<CompositeCatalogProvider> | null = null;
  let built: CompositeCatalogProvider | null = null;
  let generation = 0;
  return {
    current(): Promise<CompositeCatalogProvider> {
      const gen = generation;
      return (current ??= loadRows().then(async (rows) => {
        const composite = new CompositeCatalogProvider(
          defaultCatalogProviders(rows, addonTransport, googleBooksKey, consent),
          { cache },
        );
        if (gen === generation) built = composite;
        // The supplier's current() is already async, so the preload stays off
        // the composite's synchronous read path. Every invalidate() rebuild
        // re-preloads, so addon install/enable/disable/uninstall re-seed it.
        await preloadDiscoverCache(cache, composite);
        return composite;
      }));
    },
    peek(): CompositeCatalogProvider | null {
      return built;
    },
    invalidate(): void {
      generation += 1;
      current = null;
      built = null;
    },
  };
}

interface RoutedDetails {
  provider: CatalogProvider;
  source: CatalogSourceInfo;
}

export class CompositeCatalogProvider implements CatalogProvider {
  private readonly debounced: { search: (query: string, page: number) => Promise<PagedResult> };
  private readonly cache: DiscoverCacheStore | null;
  private readonly nowEpochSecs: () => number;
  /** Featured cache keys with a stale-while-revalidate refresh already in flight. */
  private readonly featuredRefreshes = new Set<string>();

  constructor(
    private readonly providers: CatalogProvider[] = defaultCatalogProviders(),
    options: CompositeOptions = {},
  ) {
    this.cache = options.cache ?? null;
    this.nowEpochSecs = options.nowEpochSecs ?? (() => Math.floor(Date.now() / 1000));
    this.debounced = createSearchDebouncer(
      (query, page) => this.executeSearch(query, page),
      options.debounceMs ?? DEBOUNCE_MS,
    );
  }

  /** Sources in provider order, deduped by sourceId (first occurrence wins). */
  listSources(): CatalogSourceInfo[] {
    const seen = new Set<string>();
    const out: CatalogSourceInfo[] = [];
    for (const provider of this.providers) {
      for (const source of provider.listSources()) {
        if (!seen.has(source.sourceId)) {
          seen.add(source.sourceId);
          out.push(source);
        }
      }
    }
    return out;
  }

  /**
   * True when at least one active provider opts in, so the shell can decide
   * whether any featured work is possible at all. Mirrors Android.
   */
  supportsFeatured(sort: CatalogFeaturedSort): boolean {
    return this.searchableProviders().some((p) => p.supportsFeatured(sort));
  }

  /**
   * Featured rails fan out over the providers that opt in via
   * `supportsFeatured`, then merge with the same `mergePaged` left-fold used
   * by search. Each opted-in provider reads through the Discover cache under
   * its `f:v2:{sourceId}:{sort}` key with the 6h featured TTL:
   *
   * - a fresh entry is served with zero I/O;
   * - a stale-but-resident entry is served immediately while a guarded
   *   background refresh replaces it (stale-while-revalidate);
   * - a miss fetches and caches.
   *
   * Sources that are not active at read time are never served (design A1).
   * A provider that does not opt in is never called, so its rail can only ever
   * come back empty (fail-closed) and be hidden. Each provider stays isolated:
   * a throw maps to an empty page, never failing the whole fan-out.
   */
  async featured(sort: CatalogFeaturedSort, limit: number): Promise<PagedResult> {
    if (!Number.isInteger(limit) || limit < 1) {
      throw catalogError('INVALID_PAGE', `limit must be >= 1, got ${limit}`);
    }
    const active = this.activeSourceIds();
    const pages = await Promise.all(
      this.searchableProviders()
        .filter((provider) => provider.supportsFeatured(sort))
        .map(async (provider) => {
          try {
            const hit = this.readFeaturedHit(provider, sort, active);
            if (hit) {
              if (hit.stale) this.startFeaturedRefresh(provider, sort, limit, active, hit.key);
              return hit.page;
            }
            const result = await provider.featured(sort, limit);
            this.cacheFeatured(provider, sort, result, active);
            return result;
          } catch {
            return { results: [], nextPage: null, totalCount: 0 } satisfies PagedResult;
          }
        }),
    );
    return this.mergePaged(pages, 1);
  }

  /**
   * Per-source search: exact match over the active source set, routed to the
   * single provider that owns `sourceId`, with the same page-cache read-through
   * as `search` (the thematic rail resolves through here, so its page is cached
   * too). An unknown or inactive source fails closed with an empty page — never
   * a crash, never a silent composite search. Mirrors Android.
   */
  async searchSource(sourceId: CatalogSource, query: string, page: number): Promise<PagedResult> {
    if (!Number.isInteger(page) || page < 1) {
      throw catalogError('INVALID_PAGE', `page must be >= 1, got ${page}`);
    }
    const owner = this.searchableProviders().find((provider) =>
      provider.listSources().some((source) => source.sourceId === sourceId),
    );
    if (!owner) {
      return { results: [], nextPage: null, totalCount: 0 };
    }
    const active = this.activeSourceIds();
    const cached = this.readPageHit(owner, query, page, active);
    if (cached !== null) return cached;
    const result = await owner.search(query, page);
    this.cachePage(owner, query, page, result, active);
    return result;
  }

  /**
   * Debounced entry point: only the latest burst query issues network I/O.
   * When every I/O-capable provider has a fresh cache entry, the merged page
   * returns synchronously without waiting for debounce.
   */
  search(query: string, page: number): Promise<PagedResult> {
    if (!Number.isInteger(page) || page < 1) {
      return Promise.reject(catalogError('INVALID_PAGE', `page must be >= 1, got ${page}`));
    }
    const active = this.activeSourceIds();
    const parts = this.searchableProviders().map((provider) => {
      const cached = this.readPageHit(provider, query, page, active);
      return cached !== null
        ? { cached: true as const, value: cached }
        : { cached: false as const, provider };
    });
    const needsFetch = parts.some((part) => {
      if (part.cached) return false;
      const source = singleSource(part.provider);
      return source !== null && active.has(source.sourceId);
    });
    if (!needsFetch) {
      // Only I/O-free providers (no single routable source) remain to call.
      return Promise.all(
        parts.map((part) =>
          part.cached ? Promise.resolve(part.value) : part.provider.search(query, page),
        ),
      ).then((pages) => this.mergePaged(pages, page));
    }
    return this.debounced.search(query, page);
  }

  private async executeSearch(query: string, page: number): Promise<PagedResult> {
    // Re-check inside the debounce window: a concurrent caller may have filled it.
    const active = this.activeSourceIds();
    const pages = await Promise.all(
      this.searchableProviders().map(async (provider) => {
        const cached = this.readPageHit(provider, query, page, active);
        if (cached !== null) return cached;
        const result = await provider.search(query, page);
        this.cachePage(provider, query, page, result, active);
        return result;
      }),
    );
    return this.mergePaged(pages, page);
  }

  /** Providers without sources (disabled) contribute nothing and are never called. */
  private searchableProviders(): CatalogProvider[] {
    return this.providers.filter((p) => p.listSources().length > 0);
  }

  private activeSourceIds(): Set<string> {
    const ids = new Set<string>();
    for (const provider of this.searchableProviders()) {
      for (const source of provider.listSources()) ids.add(source.sourceId);
    }
    return ids;
  }

  /** Existence check (design A1): only sources active at read time may hit. */
  private readPageHit(
    provider: CatalogProvider,
    query: string,
    page: number,
    active: Set<string>,
  ): PagedResult | null {
    if (!this.cache) return null;
    const source = singleSource(provider);
    if (!source || !active.has(source.sourceId)) return null;
    const hit = this.cache.get(pageCacheKey(source.sourceId, query, page), this.nowEpochSecs());
    return hit ? (JSON.parse(hit) as PagedResult) : null;
  }

  private cachePage(
    provider: CatalogProvider,
    query: string,
    page: number,
    result: PagedResult,
    active: Set<string>,
  ): void {
    if (!this.cache) return;
    const source = singleSource(provider);
    if (!source || !active.has(source.sourceId)) return;
    this.cache.put(
      pageCacheKey(source.sourceId, query, page),
      JSON.stringify(result),
      this.nowEpochSecs(),
      PAGE_TTL_S,
    );
  }

  /**
   * Featured read-through under `f:v2:{sourceId}:{sort}`. Returns the resident
   * page plus whether it is stale; a corrupt payload is a miss (refetch), and
   * an inactive or unroutable source is never served.
   */
  private readFeaturedHit(
    provider: CatalogProvider,
    sort: CatalogFeaturedSort,
    active: Set<string>,
  ): { key: string; page: PagedResult; stale: boolean } | null {
    if (!this.cache) return null;
    const source = singleSource(provider);
    if (!source || !active.has(source.sourceId)) return null;
    const key = featuredCacheKey(source.sourceId, sort);
    const cached = readCachedPayload(this.cache, key, this.nowEpochSecs());
    if (!cached) return null;
    try {
      return { key, page: JSON.parse(cached.payload) as PagedResult, stale: cached.stale };
    } catch {
      return null;
    }
  }

  private cacheFeatured(
    provider: CatalogProvider,
    sort: CatalogFeaturedSort,
    result: PagedResult,
    active: Set<string>,
  ): void {
    if (!this.cache) return;
    const source = singleSource(provider);
    if (!source || !active.has(source.sourceId)) return;
    this.cache.put(
      featuredCacheKey(source.sourceId, sort),
      JSON.stringify(result),
      this.nowEpochSecs(),
      FEATURED_TTL_S,
    );
  }

  /**
   * Stale-while-revalidate refresh: at most one in flight per cache key. A
   * success replaces the resident entry; a failure is swallowed so the stale
   * value already served stays the answer and the rail never depends on a
   * refresh nobody awaited.
   */
  private startFeaturedRefresh(
    provider: CatalogProvider,
    sort: CatalogFeaturedSort,
    limit: number,
    active: Set<string>,
    key: string,
  ): void {
    if (this.featuredRefreshes.has(key)) return;
    this.featuredRefreshes.add(key);
    void (async () => {
      try {
        const result = await provider.featured(sort, limit);
        this.cacheFeatured(provider, sort, result, active);
      } catch {
        // Swallowed: the served stale page remains the best answer available.
      } finally {
        this.featuredRefreshes.delete(key);
      }
    })();
  }

  /**
   * Ordered merge: left-fold the provider pages — earlier providers win fields,
   * later ones fill cover gaps and append unmatched books (the [Gutendex,
   * OpenLibrary] fold reproduces the legacy hardcoded-pair merge exactly).
   */
  private mergePaged(pages: PagedResult[], page: number): PagedResult {
    if (pages.length === 0) return toPagedResult([], page, 0);
    let results = pages[0]!.results;
    let totalCount = pages[0]!.totalCount;
    for (let i = 1; i < pages.length; i++) {
      results = mergeResults(results, pages[i]!.results);
      totalCount = resolveTotalCount(totalCount, pages[i]!.totalCount);
    }
    return toPagedResult(results, page, totalCount);
  }

  /**
   * Exact-prefix routing over the active source set: `gutendex:`/`openlibrary:`
   * book ids hit the built-ins, `addon:<addonId>:<bookId>` the owning addon.
   * Unroutable ids reject NOT_FOUND without any I/O.
   */
  async getDetails(id: string): Promise<CatalogBook> {
    const route = this.routeDetails(id);
    if (!route) {
      throw catalogError('NOT_FOUND', `unknown catalog id ${id}`);
    }
    const { provider, source } = route;
    const active = this.activeSourceIds();
    if (this.cache && active.has(source.sourceId)) {
      const hit = this.cache.get(detailCacheKey(source.sourceId, id), this.nowEpochSecs());
      if (hit) return JSON.parse(hit) as CatalogBook;
    }
    const book = await provider.getDetails(id);
    if (this.cache && active.has(source.sourceId)) {
      this.cache.put(
        detailCacheKey(source.sourceId, id),
        JSON.stringify(book),
        this.nowEpochSecs(),
        DETAIL_TTL_S,
      );
    }
    return book;
  }

  private routeDetails(id: string): RoutedDetails | null {
    let best: { provider: CatalogProvider; source: CatalogSourceInfo; prefix: string } | null =
      null;
    for (const provider of this.searchableProviders()) {
      for (const source of provider.listSources()) {
        const prefix = bookIdPrefixForSource(source.sourceId);
        if (!prefix || !id.startsWith(prefix) || id.length <= prefix.length) continue;
        if (!best || prefix.length > best.prefix.length) {
          best = { provider, source, prefix };
        }
      }
    }
    return best ? { provider: best.provider, source: best.source } : null;
  }

  resolveDownloadUrl(formats: Record<string, string>, preferEpub: boolean): string {
    return resolveDownloadUrl(formats, preferEpub);
  }

  /**
   * Addon access resolve (slice 9): route by the existing
   * `bookIdPrefixForSource` over the active source set (longest-prefix wins,
   * same as `getDetails`) and delegate to the owning provider that implements
   * `resolveAddonAccess`. No owner or no implementation ⇒ empty result.
   * `getDetails` prefix routing and `NOT_FOUND` zero-I/O are unchanged.
   */
  async resolveAddonAccess(book: CatalogBook): Promise<AddonAccessResolution> {
    const route = this.routeDetails(book.id);
    if (!route) {
      return { ...EMPTY_ADDON_ACCESS, options: [] };
    }
    const resolve = route.provider.resolveAddonAccess;
    if (typeof resolve !== 'function') {
      return { ...EMPTY_ADDON_ACCESS, options: [] };
    }
    return resolve.call(route.provider, book);
  }
}
